osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,189 @@
+                if not personal_access_token:
+                    self.failed_login = True
+                else:
+                    user.oauth_token = personal_access_token
+
+            if not self.failed_login:
+                return (yield from self._populate_user_data(user, username, password))
+            else:
+                return
+
+    @asyncio.coroutine
+    def _populate_user_data(self, user, username, password):
+        """ Takes a user object, makes a request to ensure auth is working,
+            and fills in any missing user data.
+
+            :return: user
+        """
+        me = furl.furl(settings.API_BASE)
+        me.path.add('/v2/users/me/')
+        header = {'Authorization': 'Bearer {}'.format(user.oauth_token)}
+        try:
+            resp = yield from aiohttp.request(method='GET', url=me.url, headers=header)
+        except (aiohttp.errors.ClientTimeoutError, aiohttp.errors.ClientConnectionError, aiohttp.errors.TimeoutError):
+            # No internet connection
+            self.display_error('Unable to connect to server. Check your internet connection or try again later')
+        except (aiohttp.errors.HttpBadRequest, aiohttp.errors.BadStatusLine):
+            # Invalid credentials, delete possibly revoked PAT from user data and try again.
+            # TODO: attempt to delete token from remote user data
+            user.oauth_token = yield from self._authenticate(username, password)
+            if not self.failed_login:
+                yield from self._populate_user_data(user, username, password)
+            else:
+                return
+        else:
+            json_resp = yield from resp.json()
+            remote_user = RemoteUser(json_resp['data'])
+            user.full_name = remote_user.name
+            user.osf_id = remote_user.id
+            user.username = username
+            user.password = generate_hash(password)
+
+            return user
+
+    @asyncio.coroutine
+    def log_in(self, username=None, password=None):
+        """ Takes standard auth credentials, returns authenticated user or none.
+        """
+        self.failed_login = False
+        if not username or not password:
+            self.display_error('Username and password required for login.')
+            return
+        user = yield from self._find_or_create_user(username, password)
+        if self.failed_login:
+            return
+        if user:
+            user.logged_in = True
+            try:
+                save(session, user)
+            except Exception as e:
+                self.display_error('Unable to save user data. Please try again later', e=e)
+                return
+            else:
+                return user
+
+        return
+
+    def display_error(self, message, fail=True, e=None):
+        """ Displays error message to user, sets fail flag unless otherwise specified,
+            and logs exception if given.
+        """
+        if e:
+            logging.exception(e)
```

Comment:
See note above about logging.exception, which should only be called from within an exception handler.
https://docs.python.org/3/library/logging.html#logging.Logger.exception

You can log any exception object in a standard `logging.<level>()` call (outside the exception handler) by passing in the `exc_info` argument:
https://docs.python.org/3/library/logging.html#logging.Logger.debug